Objects/listobject.c

```diff
@@ -3444,18 +3444,23 @@ listiter_reduce_general(void *_it, int forward)
 {
     PyObject *list;
 
+    /* _PyEval_GetBuiltin can invoke arbitrary code.
+     * calls must be *before* access of `_it` pointers,
+     * since C parameter eval order is undefined.
+     * see issue #101765 */
+
     /* the objects are not the same, index is of different types! */
     if (forward) {
+        PyObject *iter = _PyEval_GetBuiltin(&_Py_ID(iter));
         _PyListIterObject *it = (_PyListIterObject *)_it;
         if (it->it_seq) {
-            return Py_BuildValue("N(O)n", _PyEval_GetBuiltin(&_Py_ID(iter)),
-                                 it->it_seq, it->it_index);
+            return Py_BuildValue("N(O)n", iter, it->it_seq, it->it_index);
         }
     } else {
+        PyObject *reversed = _PyEval_GetBuiltin(&_Py_ID(reversed));
```

Comment:
is it worth also testing the `reversed` case for lists?

Reply:
Added case for this in [49ba8c3](https://github.com/python/cpython/pull/101769/commits/49ba8c3bfc09e026970bc4fa842c466dcee7d0f5)